src/Microsoft.AspNetCore.Http/PipeWriterAdapter.cs

```diff
@@ -0,0 +1,287 @@
+        public override Span<byte> GetSpan(int sizeHint = 0)
+        {
+            EnsureCapacity(sizeHint);
+
+            return _currentSegment.AsSpan(_position, _currentSegment.Length - _position);
+        }
+
+        /// <inheritdoc />
+        public override void CancelPendingFlush()
+        {
+            _currentFlushCanceled = true;
+            InternalTokenSource?.Cancel();
+        }
+
+        /// <inheritdoc />
+        public override void Complete(Exception exception = null)
+        {
+            _isCompleted = true;
+            if (exception != null)
+            {
+                _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+            }
+        }
+
+        /// <inheritdoc />
+        public override void OnReaderCompleted(Action<Exception, object> callback, object state)
+        {
+            if (callback == null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            // Note this currently doesn't do anything.
+            // Implementing completions/callbacks would require creating a PipeReaderAdapter.
+        }
+
+        /// <inheritdoc />
+        public override ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = default)
+        {
+            return FlushAsyncInternal(cancellationToken);
+        }
+
+        private async ValueTask<FlushResult> FlushAsyncInternal(CancellationToken cancellationToken = default)
+        {
+            if (_bytesWritten == 0)
+            {
+                return new FlushResult(isCanceled: false, IsCompletedOrThrow());
+            }
+
+            // Wrap the provided cancellationToken with an internal one
+            // to allow CancelPendingFlush to cancel writes and flushes
+            if (_currentFlushCanceled)
+            {
+                // If CancelPendingFlush was already called, we need to return a FlushResult with
+                // Canceled. At this point, we create a fresh CTS that isn't canceled.
+                var result = new FlushResult(isCanceled: true, IsCompletedOrThrow());
+                InternalTokenSource = new CancellationTokenSource();
+                _currentFlushCanceled = false;
+                return result;
+            }
+
+            // As an optimization, we only create a new CTS if the previous one wasn't canceled.
+            // Otherwise, we would need to create a new CTS for every call to FlushAsync.
+            var joinedToken = CancellationTokenSource.CreateLinkedTokenSource(
+                cancellationToken,
+                InternalTokenSource.Token).Token;
+
+            try
+            {
+                // Write all completed segments and whatever remains in the current segment
+                // and flush the result.
+                if (_completedSegments != null)
+                {
+                    var count = _completedSegments.Count;
+                    for (var i = 0; i < count; i++)
```

Comment:
foreach?

Reply:
A long while ago, I remember for was more preformant than foreach.